Fix template annotations

ADmad opened the pull request:
<!---

Please describe the reason for the changes you're proposing. If it is a large change, please open an issue to discuss first.

Always follow the [contribution guidelines](https://github.com/cakephp/cakephp/blob/master/.github/CONTRIBUTING.md) when submitting a pull request. In particular, make sure existing tests still pass, and add tests for all new behavior. When fixing a bug, you may want to add a test to verify the fix.

-->

ADmad committed:
Fix template annotations

dereuromark committed:
Add inline @phpstan-ignore for defensive code patterns

Replace baseline entries with inline ignores for:
- isset.property checks on resources in destructors
- method_exists() checks for patch method on EntityInterface
- is_callable() checks after method_exists()
- SASL support detection in MemcachedEngine

Also removes incorrect @var annotations that were less specific
than PHPStan's inferred types.

dereuromark committed:
Fix flaky test due to Date header timing

dereuromark commented:
A few suggestions that could hep improve this further:

 1. Add template types to constructor @param docblocks

  Currently the templates are defined but not bound. To enable actual type inference:
```php
  // Before (PR #19118):
  /**
   * @template TKey
   * @template TValue
   * @extends Collection<TKey, TValue>
   */
  class BufferedIterator extends Collection
  {
      /**
       * @param iterable $items The items to be filtered.
       */
      public function __construct(iterable $items)

  // Better:
  /**
   * @template TKey
   * @template TValue
   * @extends Collection<TKey, TValue>
   */
  class BufferedIterator extends Collection
  {
      /**
       * @param iterable<TKey, TValue> $items The items to be filtered.
       */
      public function __construct(iterable $items)
```
  This lets PHPStan infer types when you create iterators:
  $users = new BufferedIterator($userEntities); // PHPStan knows TValue is User

  2. Fix ZipIterator with inline ignore instead of baseline

  Instead of adding a new baseline entry, use:
  /** @phpstan-ignore argument.type */
  count($value)

  Keeps the baseline shrinking, not growing.

  3. Fix Collection base class too

  The parent Collection::__construct() also has untyped @param iterable $items. Should be @param iterable<TKey, TValue> $items.

  4. Use @template-covariant for TValue on read-only types

  For ResultSetInterface and similar read-only containers:
  @template TKey
  @template-covariant TValue

  This allows ResultSetInterface<int, User> to be used where ResultSetInterface<int, EntityInterface> is expected.

dereuromark committed:
Improve template type inference for Collection classes

- Add @param iterable<TKey, TValue> to Collection and iterator constructors
  This enables PHPStan to infer template types from constructor arguments

- Replace ZipIterator baseline entry with inline @phpstan-ignore
  Keeps the baseline cleaner while still suppressing expected errors

dereuromark commented:
Pushed improvements to make the template annotations more useful:

**1. Added `@param iterable<TKey, TValue>` to constructors**

Collection and all iterator classes now have typed constructor parameters:
```php
/**
 * @param iterable<TKey, TValue> $items Items.
 */
public function __construct(iterable $items)
```

This enables PHPStan to infer template types from constructor arguments:
```php
$users = new BufferedIterator($userEntities); 
// PHPStan now knows TValue is User
```

**2. Replaced baseline entry with inline ignore**

The ZipIterator count() issue is now handled with an inline `@phpstan-ignore` in CollectionTrait, removing one entry from the baseline. Net effect: baseline shrinks by 1 more entry.

Files changed:
- `Collection.php` - constructor param type
- `CollectionTrait.php` - inline phpstan-ignore for cartesianProduct
- 10 iterator classes - constructor param types
- `phpstan-baseline.neon` - removed ZipIterator entry

ADmad committed:
Merge pull request #19119 from cakephp/template-annotations-2

Add inline @phpstan-ignore for defensive code patterns

dereuromark committed:
Fix PHPStan errors from typed constructor param

Add inline ignores for cases where array types don't match
template parameters:
- Collection::__unserialize() passes plain array
- SortIterator passes sorted array with transformed keys
- CollectionTrait::transpose() calls count(current()) on TValue

dereuromark committed:
Remove baseline entry now handled by inline ignore

dereuromark commented:
@rochamarcelo You also OK with all of those?

rochamarcelo commented:
Looks good, but what is the use of annotations on  ExtractIterator, wont this result in wrong type since the final data has different type than the original data? NestIterator also results in different data type

dereuromark commented:
>  This enables PHPStan to infer template types from constructor arguments

is that not true?

rochamarcelo commented:
This would cause error because phpstan understand $names is a collection of Entity not a collection of string ('author.name')
```
$names = new ExtractIterator($items, 'author.name');
foreach ($names as $name) {
      echo strtoupper($name);
}
```

rochamarcelo commented:
Same for NestIterator and ZipIterator , the result type is not the same as initial type (the one used on constructor)

dereuromark commented:
Ah ok

  Iterators that transform output type (need to revert):
  - ExtractIterator - extracts a property path, output is the extracted value type (not the original item)
  - ReplaceIterator - callback transforms items, output could be any type
  - ZipIterator - combines iterators, output is arrays or callback result

  Iterators that preserve type (OK to keep):
  - BufferedIterator, FilterIterator, InsertIterator, NestIterator, NoChildrenIterator, SortIterator, StoppableIterator, UniqueIterator - all preserve the item type

  Let me revert the annotations on the three transforming iterators.

And:
  - NestIterator: Input T[], output T[] (same type, just recursive traversal)

should be fine to keep?

dereuromark committed:
Revert template annotations on type-transforming iterators

ExtractIterator, ReplaceIterator, and ZipIterator transform
their output types, so propagating TValue from input to output
is incorrect and causes PHPStan false positives.

rochamarcelo commented:
For NestIterator is propably tricky to get the correcty type, the initial type is something like `Collection<int, array<int, Entity>>` and the result type is Collection<int, Entity>

dereuromark committed:
Revert NestIterator template annotation

Children extracted via nestKey can be a different type than parent
items, so the type-preserving annotation is incorrect.

markstory commented on src/Collection/Iterator/InsertIterator.php:
```diff
@@ -59,11 +63,11 @@ class InsertIterator extends Collection
      * @param string $path A dot separated list of properties that need to be traversed
      * to insert the value into the target collection.
-     * @param iterable $values The source collection from which the values will
+     * @param iterable<mixed, mixed> $values The source collection from which the values will
```
Shouldn't this be the templated value? This is the input value types.

dereuromark replied:
I thought <mixed, mixed> is correct here - the inserted values are intentionally a different type from TValue. Using <TKey, TValue> would be wrong because it would claim the values have the same type as the items.

Adding a third template TInsertValue adds complexity with no practical benefit since current() returns mixed anyway

ADmad commented:
These are the remaining template params related errors reported by psalm (that's how I discovered them initially)

```
ERROR: MissingTemplateParam - src/Collection/Iterator/ExtractIterator.php:28:7 - Cake\Collection\Iterator\ExtractIterator has missing template params when extending Cake\Collection\Collection, expecting 2 (see https://psalm.dev/182)
class ExtractIterator extends Collection


ERROR: MissingTemplateParam - src/Collection/Iterator/NestIterator.php:29:7 - Cake\Collection\Iterator\NestIterator has missing template params when extending Cake\Collection\Collection, expecting 2 (see https://psalm.dev/182)
class NestIterator extends Collection implements RecursiveIterator


ERROR: MissingTemplateParam - src/Collection/Iterator/ReplaceIterator.php:29:7 - Cake\Collection\Iterator\ReplaceIterator has missing template params when extending Cake\Collection\Collection, expecting 2 (see https://psalm.dev/182)
class ReplaceIterator extends Collection


ERROR: MissingTemplateParam - src/Collection/Iterator/TreeIterator.php:30:65 - Cake\Collection\Iterator\TreeIterator has missing template params when extending Cake\Collection\CollectionInterface, expecting 2 (see https://psalm.dev/182)
class TreeIterator extends RecursiveIteratorIterator implements CollectionInterface


ERROR: MissingTemplateParam - src/Collection/Iterator/TreePrinter.php:30:64 - Cake\Collection\Iterator\TreePrinter has missing template params when extending Cake\Collection\CollectionInterface, expecting 2 (see https://psalm.dev/182)
class TreePrinter extends RecursiveIteratorIterator implements CollectionInterface


ERROR: MissingTemplateParam - src/Collection/Iterator/ZipIterator.php:46:30 - Cake\Collection\Iterator\ZipIterator has missing template params when extending Cake\Collection\CollectionInterface, expecting 2 (see https://psalm.dev/182)
class ZipIterator implements CollectionInterface
```

dereuromark commented:
Should we keep as is?

Or add
```
  // For ExtractIterator, ReplaceIterator, etc:
  @extends \Cake\Collection\Collection<mixed, mixed>

  // For ZipIterator:
  @implements \Cake\Collection\CollectionInterface<mixed, mixed>
```
etc?

ADmad commented:
For `ExtractIterator` and `ReplaceIterator` it should be `@extends \Cake\Collection\Collection<TKey, mixed>`. Pretty sure only the values gets changed, the key remains the same.

dereuromark committed:
Add template annotations to type-transforming iterators

- ExtractIterator, ReplaceIterator: Use <TKey, mixed> since keys are
  preserved but values are transformed
- ZipIterator: Use <mixed, mixed> since both keys and values change
- NestIterator: Add @extends Collection<mixed, mixed>

dereuromark commented:
@ADmad I let you cover the remaining things I might have missed.

ADmad commented:
This is good enough for now. Can do another pass in the future.

ADmad commented:
@rochamarcelo Does this seems good enough now?

rochamarcelo commented:
Good to go.
